pkg/crawl/identify.go

```diff
@@ -106,99 +107,159 @@ type Log4jIdentifier struct {
+
+	paths := []string{path}
+	log4jNameMatch, nameVulnerability, nameVersions := i.findArchiveNameVulnerability(paths)
+	inArchive, inZipVs, skipped, err := i.findArchiveContentVulnerabilities(ctx, 0, walker, obfuscated, paths)
 	if err != nil {
-		return 0, nil, 0, errors.Wrapf(err, "failed to walk archive %s", path)
+		return 0, errors.Wrapf(err, "failed to walk archive %s", path)
+	}
+
+	reportFindings, reportVersions := resolveNameAndContentFindings(log4jNameMatch, nameVulnerability, nameVersions, inArchive, inZipVs)
+	if reportFindings == NothingDetected {
+		return skipped, err
+	}
+
+	if obfuscated {
+		// make sure this is in the right place
+		i.printInfoFinding("Found finding in what appeared to be an obfuscated jar at %s", path)
+		reportFindings |= JarFileObfuscated
+	}
+
+	i.Report(ctx, path, reportFindings, reportVersions)
+	return skipped, err
+}
+
+func resolveNameAndContentFindings(nameMatchesLog4jJar bool, nameFinding Finding, nameVersions Versions, archiveFinding Finding, archiveVersions Versions) (Finding, Versions) {
+	if archiveFinding == NothingDetected {
+		if nameFinding != NothingDetected {
+			return nameFinding, nameVersions
+		}
+		return NothingDetected, nil
+	}
+
+	// If detections are made but no clues of any vulnerable versions within the archive,
+	// then we determine non-vulnerable if the archive name has been matched but with a
+	// non-vulnerable filename.
+	if len(archiveVersions) == 0 && nameMatchesLog4jJar && nameFinding == NothingDetected {
+		return NothingDetected, nil
 	}
-	for v := range inZipVs {
+
+	findings := nameFinding | archiveFinding
+	if findings == NothingDetected {
+		return findings, nil
+	}
+
+	versions := archiveVersions
+	for v := range nameVersions {
 		versions[v] = struct{}{}
 	}
-	// If file on disk matches log4j but no signs of vulnerable version have been found
-	// during identification phase, then we assume non-vulnerable.
-	if log4jMatch && len(versions) == 0 {
-		return NothingDetected, nil, skipped, nil
+	if len(versions) == 0 {
+		versions = map[string]struct{}{UnknownVersion: {}}
 	}
-	if inZip != NothingDetected && obfuscated {
-		result |= JarFileObfuscated
-		i.printInfoFinding("Found finding in what appeared to be an obfuscated jar at %s", path)
+	return findings, versions
+}
+
+func (i *Log4jIdentifier) findArchiveNameVulnerability(nestedPaths nestedPaths) (bool, Finding, Versions) {
+	path := nestedPaths[len(nestedPaths)-1]
+	var jarNameMatch bool
+	var jarVersion string
+	var jarFinding Finding
+	if len(nestedPaths) == 1 {
+		// we are on disk and so need to split using filepath to support different OS separators
+		_, filename := filepath.Split(path)
+		jarVersion, jarNameMatch = FileNameMatchesLog4jJar(filename)
+		jarFinding = JarName
+	} else {
+		// we are in an archive and so can assume separator is '/'
+		jarVersion, jarNameMatch = pathMatchesLog4JVersion(path)
+		jarFinding = JarNameInsideArchive
+		if jarNameMatch {
```

Comment:
This behaviour is carried over from previous implementation.

@hpryce do we want to log at info level when we match _any_ `log4j-core` jar or should we only do it when vulnerable?